### ISSUE TITLE
Incorrect caching of skipif/xfail string condition evaluation

### ISSUE THREAD

**Issue by bluetech** [pytest-dev/pytest#7360]
Version: pytest 5.4.3, current master

pytest caches the evaluation of the string in e.g. `@pytest.mark.skipif("sys.platform == 'win32'")`. The caching key is only the string itself (see `cached_eval` in `_pytest/mark/evaluate.py`). However, the evaluation also depends on the item's globals, so the caching can lead to incorrect results. Example:

```py
# test_module_1.py
import pytest

skip = True

@pytest.mark.skipif("skip")
def test_should_skip():
    assert False
```

```py
# test_module_2.py
import pytest

skip = False

@pytest.mark.skipif("skip")
def test_should_not_skip():
    assert False
```

Running `pytest test_module_1.py test_module_2.py`.

Expected: `test_should_skip` is skipped, `test_should_not_skip` is not skipped.

Actual: both are skipped.

---

I think the most appropriate fix is to simply remove the caching, which I don't think is necessary really, and inline `cached_eval` into `MarkEvaluator._istrue`.

**Comment by nicoddemus** [comment]
> I think the most appropriate fix is to simply remove the caching, which I don't think is necessary really, and inline cached_eval into MarkEvaluator._istrue.

I agree:

* While it might have some performance impact with very large test suites which use marks with eval, the simple workaround is to not use the eval feature on those, which is more predictable anyway.
* I don't see a clean way to turn "globals" in some kind of cache key without having some performance impact and/or adverse effects.

So 👍 from me to simply removing this caching.

**Comment by RonnyPfannschmidt** [comment]
As globals are dynamic, i would propose to drop the cache as well, we should investigate reinstating a cache later on

**Comment by bluetech** [comment]
Fixed in #7373
